### DISCUSSION TITLE
DKMaterial

### DISCUSSION THREAD

**Discussion by arkiny** [Hongtae/DKGL#13] [Q&A] [answered]
@Hongtae 

작업하면서 DKMaterial 인터페이스(아직 작업이 덜되보이는...)에 따라서 
일단 샘플코드 내에서 한번 랩핑한 클래스에서 퍼블릭으로 열어넣은 변수들에 값들을 우겨 넣을 예정인데요,

혹시 DKMaterial 관련해서 생각해두신 방식이 있으신가요?

Mesh를 Wrapping하는 코드는 아래 코드를 참고하시면 될 듯합니다.

https://github.com/DKGL/DKGL_Samples/blob/ade1e2a781cc4e3514354befc6d270dd1c008dfe/Samples/DKGLDemoScene/StaticMesh.cpp#L193

DKGL에 실제 인테그레이션 하기 전에는 위와 같이 샘플코드에서 한번 랩핑해서 테스트하고, 
검토받고 프레임워크에 인테그레이션 할 예정입니다.

**Comment by arkiny** [comment]
추가적으로 현재 DKMaterial의 경우 Shader Data 파일 하나당 하나의 셰이더 펑션을 처리하게 되어있는데, vertex shader와 fragment shader가 나뉘어져 있는 경우, DKMaterial안에서 처리하는 DKRenderPipelineDescriptor가 동작을 못할것 같아요. 일단 Map을 사용하도록 변경해서 사용하도록 하겠습니다. 

`class DKGL_API DKMaterial : public DKResource
    {
    public:
        DKMaterial();
        ~DKMaterial();

        struct ShaderTemplate
        {
            DKObject<DKShader> shader;
            DKObject<DKShaderModule> shaderModule;

            struct FunctionProperty
            {
                DKArray<DKShaderBindingSetLayout> bindingSetLayouts;
                DKArray<DKShaderAttribute> inputAttributes;
            };

            DKMap<DKString, FunctionProperty> functionProperties;
            DKMap<DKString, DKShaderUniform> predefinedUniformValues;
        };

         //DKObject<ShaderTemplate> shaderTemplate; 수정전
        DKMap<DKShaderStage, DKObject<ShaderTemplate>> shaderTemplate;

        //DKObject<DKShaderFunction> shaderFunction; 수정전
        DKMap<DKShaderStage, DKObject<DKShaderFunction>> shaderFunction;
        DKObject<DKRenderPipelineState> renderPipelineState;
`

**Comment by arkiny** [comment]
DKMaterial 안에 renderPipelineDecription이 있는데, 렌더타겟 관련 설정도 해당 디스크립션에서 변경이 가능한걸로 보아 머티리얼(재질)과는  크게 관계 있어보이진 않습니다...

**Reply by arkiny** [reply]
이것도 일단은 swapchain에 마지막에 옮길때의 문제인거라, 별도로 한번 커맨드를 쪼개서 처리하면 되니 이 문제는 무시하셔도 될듯합니다.

**Reply by Hongtae** [reply]
이건 아마 pso 를 빌드하기 위해 있는 데이터인것 같습니다. 디스크에서 시리얼라이즈 된 상태에서는 pso 는 없고, 그것을 (나중에) 생성하기 위한 descriptor 만 존재할거 같네요.. (그래서 renderPipelineDescriptor 가 함께 있는듯)

**Reply by Hongtae** [reply]
아시겠지만, descriptor 에 있는 렌더타겟 (colorAttachments, depthStencilDescriptor) 정보는 렌더타겟 규격일 뿐이라, 실제 렌더타겟이 무엇으로 바인딩 되는지는 RenderCommandEncoder 에서 결정할 수 있습니다.

**Reply by arkiny** [reply]
넵, 그래서 앞으로 샘플 작업하면서 차차 HDR to LDR 및 ToneMapping 처리하면서 SwapChain용으로  파이프라인을 하나 더 만들까 생각중입니다. 일단은 아래 코드처럼만 처리해두었고 좀더 확장될거에요 👍

**Comment by arkiny** [comment, marked as answer]
만약 stage별로 function과 shaderTemplate을 쪼개게 된다면 링크에 있는 코드처럼 진행될듯합니다.
https://github.com/DKGL/DKGL_Samples/blob/7f9c37ba96b3692bb2e589ccf9fd52cb93f37d15/Samples/DKGLDemoScene/StaticMesh.cpp#L167

**Reply by Hongtae** [reply]
오 멋집니다. 잘 돌아가나요?
말씀하신대로 shaderTemplate 은 stage 별로 있는게 맞고..  (이건 쉐이더 리소스를 바인딩하는 레이아웃 템플릿으로 보입니다.)
파이프라인에 사용된 shaderFunction 도 stage 별로 있는게 맞는것 같습니다.
왜 저렇게 되어있는지는 잘 기억이 안나네요.. @_@

**Reply by arkiny** [reply]
네 저렇게 작업하니 잘 돌아갑니다. 
작업 하면서 괜찮은 부분에 있어선 DKMaterial에도 반영해두도록 하겠습니다

**Reply by Hongtae** [reply]
저부분은 한글주석만 빼시고 변수명을 복수형으로 하심 딱 좋아보이네요.

**Reply by arkiny** [reply]
맨위에 코드는 그냥 예시로 올려둔거라 ㅎㅎ  말씀하신대로 진행해보겠습니다.

**Comment by Hongtae** [comment]
이거 수정된 버전이 merge 되었습니다.
역할은, 추상적으로는 비주얼에 대한 제어이고, 시스템상(구현상) 으로는 파이프라인에 필요한 모든 리소스를 자동으로 바인딩 해주는 일 입니다.

사용자는 material 에는 기본데이터 (템플릿)을 넣어두고 외부에 인스턴스(실제 바인딩 될 값) 으로 바인딩 할수 있습니다.
버퍼를 통째로 바인딩 하거나, 다른 버퍼 와 합쳐서 바인딩할수 있고, 버퍼 의 컴포넌트(Struct-Element)를 각각 따로 바인딩 할수 있습니다.
그리고 실제 mutable 한 값들은 mesh 안에 보관하는것을 권장하는데, 이부분은 material-instance 같은것으로 구체화 하면 좋을것 같습니다만, 스펙이 크지 않아서 (단순 데이터 저장) 그냥 mesh 안에 두었네요. ㅋ
그리고 MRT, Depth 버퍼 정보등, RenderPass 관련된것도 아직 mesh 안에 그냥 있습니다.. =_=;; 
이거 RenderPass 와 연동하려 하니, PSO 다시 만드는것도 그렇고, 나중에 에디터에서 에셋 뽑아낼때는 바뀔일이 별로 없을것 같기도 합니다..(??)
이부분은 Canvas 에 적용되면서 조금 더 세분화 될것 같습니다. 

구현 목표는 User-Defined 를 제외한 모든것은 자동 바인딩이지만 Scene (DKSceneState) 작업이 아직 구현이 덜되서 이번에 들어가지 못했습니다.

**Reply by Hongtae** [reply]
DKGL_Sample 에도 업데이트 되어있습니다. (Material 샘플)
복잡도는 이전것과 큰 차이는 없네요.. ㅠㅠ (바인딩 정보를 찾아서 매칭해주는것에 의미를..)

**Reply by arkiny** [reply]
오랜만에 메이저 업데이트네요 😄 
찬찬히 확인해보고 진행해보겠습니다.
